### PR TITLE
Fix data race of Token result Pool

### PR DESCRIPTION
**Pull request by louyuting**
### Describe what this PR does / why we need it
```
==================
WARNING: DATA RACE
Read at 0x00c0000a0a00 by goroutine 11:
  github.com/alibaba/sentinel-golang/core/base.(*TokenResult).IsBlocked()
      /Users/sczyh30/temp/to-review/sentinel-go-ximu/core/base/result.go:58 +0x351
  github.com/alibaba/sentinel-golang/core/stat.(*StatisticSlot).OnCompleted()
      /Users/sczyh30/temp/to-review/sentinel-go-ximu/core/stat/stat_slot.go:32 +0x40
  github.com/alibaba/sentinel-golang/core/base.(*SlotChain).exit()
      /Users/sczyh30/temp/to-review/sentinel-go-ximu/core/base/slot_chain.go:181 +0x9f
  github.com/alibaba/sentinel-golang/core/base.(*SentinelEntry).Exit.func1()
      /Users/sczyh30/temp/to-review/sentinel-go-ximu/core/base/entry.go:64 +0xa3
  sync.(*Once).doSlow()
      /usr/local/go/src/sync/once.go:66 +0x103
  sync.(*Once).Do()
      /usr/local/go/src/sync/once.go:57 +0x68
  github.com/alibaba/sentinel-golang/core/base.(*SentinelEntry).Exit()
      /Users/sczyh30/temp/to-review/sentinel-go-ximu/core/base/entry.go:62 +0x144
  main.main.func2()
      /Users/sczyh30/temp/to-review/sentinel-go-ximu/api/test/cb.go:67 +0x136

Previous write at 0x00c0000a0a00 by goroutine 10:
  github.com/alibaba/sentinel-golang/core/base.RefurbishTokenResult()
      /Users/sczyh30/temp/to-review/sentinel-go-ximu/core/base/result.go:118 +0x40a
  github.com/alibaba/sentinel-golang/core/base.(*SlotChain).Entry()
      /Users/sczyh30/temp/to-review/sentinel-go-ximu/core/base/slot_chain.go:145 +0x401
  github.com/alibaba/sentinel-golang/api.entry()
      /Users/sczyh30/temp/to-review/sentinel-go-ximu/api/api.go:114 +0x4c1
  github.com/alibaba/sentinel-golang/api.Entry()
      /Users/sczyh30/temp/to-review/sentinel-go-ximu/api/api.go:86 +0x19a
  main.main.func1()
      /Users/sczyh30/temp/to-review/sentinel-go-ximu/api/test/cb.go:47 +0x58

Goroutine 11 (running) created at:
  main.main()
      /Users/sczyh30/temp/to-review/sentinel-go-ximu/api/test/cb.go:58 +0x439

Goroutine 10 (running) created at:
  main.main()
      /Users/sczyh30/temp/to-review/sentinel-go-ximu/api/test/cb.go:45 +0x421
==================
```

Root cause is here:
Refer to: [https://github.com/alibaba/sentinel-golang/commit/975debd7075d071473fcbcd17c3dbc3bd201a2da#diff-604f4192deea5adf9c159e5f88f64630R93
](https://github.com/alibaba/sentinel-golang/commit/975debd7075d071473fcbcd17c3dbc3bd201a2da#diff-604f4192deea5adf9c159e5f88f64630R93
)
```

defer func() {
	if r != nil {
		base.RefurbishTokenResult(r)
	}
}()
```
r is the same Pointer with context.Output.LastResult. Sentinel refurbish the r. But When user call Exit() by manually, will call context.Output.LastResult(the refurbished TokenResult might be got by other goroutine), it causes DATA RACE.

### Special notes for reviews

### CODE REVIEW

File: core/flow/slot.go [hunk 1]
```diff
@@ -25,20 +26,23 @@ func (s *FlowSlot) Check(ctx *base.EntryContext) *base.TokenResult {
 			continue
 		}
 		r := canPassCheck(tc, ctx.StatNode, ctx.Input.AcquireCount)
+		if r == nil {
+			// nil means pass
+			continue
+		}
 		if r.Status() == base.ResultStatusBlocked {
-			return r
+			result.DeepCopyFrom(r)
```
Comment (sczyh30) [inline core/flow/slot.go:34]: Why a copy is needed here?

File: core/system/slot.go [hunk 2]
```diff
@@ -12,17 +12,19 @@ type SystemAdaptiveSlot struct {
...
 	}
 	rules := GetRules()
+	result := ctx.RuleCheckResult
 	for _, rule := range rules {
 		passed, m := s.doCheckRule(rule)
 		if passed {
 			continue
 		}
-		return base.NewTokenResultBlockedWithCause(base.BlockTypeSystemFlow, base.BlockTypeSystemFlow.String(), rule, m)
+		result.ResetToBlockedWithCauseFrom(base.BlockTypeSystemFlow, base.BlockTypeSystemFlow.String(), rule, m)
```
Comment (sczyh30) [inline core/system/slot.go:24]: Check result nil here? (in case previous customized slots cleared the result)

File: core/flow/slot.go (hunk 1 above)
Comment (louyuting) [reply core/flow/slot.go]: It's not needed, maybe return r is enough

File: core/system/slot.go (hunk 2 above)
Comment (louyuting) [reply core/system/slot.go]: make sense

**Comment by sczyh30** [comment]
Thanks!

**Pull request merged**